### PR TITLE
feat: span transaction with left payment token time

### PR DESCRIPTION
**Pull request by pietro-tota**
Depends on https://github.com/pagopa/pagopa-ecommerce-commons/pull/88

#### List of Changes

1. Add custom elk span creation during new transaction  processing.

This span is created programmatically only when a transaction is cached for an already activated payment token.
Into custom span are present two labels:

- paymentToken --> valued with transaction payment token
- paymentTokenLeftTimeSec -> numeric label valued with payment token validity left time in seconds 

2. Move PaymentRequestInfo save after Nodo successful payment notice activation 

#### Motivation and Context
1. Those modifications are needed in order to track transactions activated close to the Nodo payment token expiration
2. Previously PaymentRequestInfo object were saved for every activation, also for those with a cache hit, causing expiration to be updated for every cache hit. With this modification Redis cache is upgraded only after Nodo activatePaymentNotice response and left as is for every subsequent cache hit, making cache leave at least as payment token duration after successful Nodo activation

#### How Has This Been Tested?
Test in DEV environment

#### Screenshots (if appropriate):
Example dashboard with sample query that take into account all subsequent activation for a given payment token
Screenshot 2023-05-31 alle 16 37 17

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by Fernando-Granato — APPROVED** [review]

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java [hunk 1]
```diff
@@ -230,6 +222,34 @@ public Mono<Tuple2<Mono<TransactionActivatedEvent>, String>> handle(
         );
     }
 
+    private void spanCustomTransaction(PaymentRequestInfo paymentRequestInfo) {
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java:225]:
```suggestion
    private void traceRepeatedActivation(PaymentRequestInfo paymentRequestInfo) {
```

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java [hunk 2]
```diff
@@ -230,6 +222,34 @@ public Mono<Tuple2<Mono<TransactionActivatedEvent>, String>> handle(
...
+        String transactionActivationDateString = paymentRequestInfo.activationDate();
+        Duration paymentTokenValidityTimeLeft = Duration.ofSeconds(paymentTokenTimeout);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java:227]: Isn't this the the payment token validity time? (i.e. should be the same as the first activation, not less)

Comment (pietro-tota) [reply src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java:227]: The paymentTokenTimeout, valued as class attribute, contains the `payment.token.validity` configured value

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java (hunk 1 above)
Comment (pietro-tota) [reply src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java]: Fixed with https://github.com/pagopa/pagopa-ecommerce-transactions-service/pull/240/commits/0da79d3734a91df7b44ea40e704686c6bfe6e30e

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java (hunk 2 above)
Comment (giovanniberti) [reply src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java:227]: Ok, I would remove `left` in the name then, what do you think?

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java [hunk 3]
```diff
@@ -230,6 +222,45 @@ public Mono<Tuple2<Mono<TransactionActivatedEvent>, String>> handle(
...
+                );
+            } else {
+                log.error(
+                        "Cannot trace repeated transaction activation for {} with payment token: {}, missing transaction activation date",
+                        paymentRequestInfo.id(),
+                        paymentRequestInfo.paymentToken()
+                );
+                span.captureException(
+                        new IllegalArgumentException(
+                                "Invalid transaction activation date null for rptId: %s"
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionActivateHandler.java:253]:
```suggestion
                                "Null transaction activation date for rptId %s in repeated activation"
```

**Review by giovanniberti — APPROVED** [review]
🚀

**Pull request merged**